In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import numpy as np
import pandas as pd

In [3]:
pd.set_option('display.max_columns', None)

# Data Extraction

In [4]:
path = "/content/drive/MyDrive/DelayedFlights.csv"
df = pd.read_csv(path)
df

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
0         2211.0        2225            WN        335  N712SW   
1         1002.0        1000            WN       3231  N772SW   
2          804.0         750            WN        448  N428WN   
3         1959.0        1925            WN       3920  N464WN   
4         2121.0        2110            WN        378  N726SW   
...          ...         ...           ...        ...     ...   
1936753   1617.0        1552            DL       1621  N938DL   
1936754    904.0         749            DL       1631  N3743H   
1936755   1149.0        1010            DL       1631  N909DA   
1936756   1446.0        1437            DL       1639  N646DL   
1936757   1413.0        1418            DL       1641  N908DL   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
0                    128.0           150.0    116.0     -14.0       8.0   
1                    128.0           145.0    113.0       2.0      19.0   
2                     96.0            90.0     76.0      14.0       8.0   
3                     90.0            90.0     77.0      34.0      34.0   
4                    101.0           115.0     87.0      11.0      25.0   
...                    ...             ...      ...       ...       ...   
1936753              147.0           152.0    120.0      25.0      30.0   
1936754              127.0           109.0     78.0      75.0      57.0   
1936755              162.0           143.0    122.0      99.0      80.0   
1936756              115.0           117.0     89.0       9.0      11.0   
1936757              123.0           135.0    104.0      -5.0       7.0   

        Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  \
0          IAD  TPA       810     4.0      8.0          0                N   
1          IAD  TPA       810     5.0     10.0          0                N   
2          IND  BWI       515     3.0     17.0          0                N   
3          IND  BWI       515     3.0     10.0          0                N   
4          IND  JAX       688     4.0     10.0          0                N   
...        ...  ...       ...     ...      ...        ...              ...   
1936753    MSP  ATL       906     9.0     18.0          0                N   
1936754    RIC  ATL       481    15.0     34.0          0                N   
1936755    ATL  IAH       689     8.0     32.0          0                N   
1936756    IAD  ATL       533    13.0     13.0          0                N   
1936757    SAT  ATL       874     8.0     11.0          0                N   

         Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0               0           NaN           NaN       NaN            NaN   
1               0           NaN           NaN       NaN            NaN   
2               0           NaN           NaN       NaN            NaN   
3               0           2.0           0.0       0.0            0.0   
4               0           NaN           NaN       NaN    

In [5]:
# Print a Series with the data type of each column
print(df.dtypes)

Unnamed: 0             int64
Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn               float64
TaxiOut              float64
Cancelled              int64
CancellationCode      object
Diverted               int64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object


# Data Cleaning
1. Extracting metadata
2. Handling Null Values
3. Ensuring certain columns aren't redundant
4. Identifying Columns with Repeating Values and Ensuring they Make Sense Considering the Context


## 1. Extracting Metadata

In [6]:
# Dropping "Unnamed" column
df = df.drop(columns=['Unnamed: 0'])
df

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0        2008      1           3          4   2003.0        1955   2211.0   
1        2008      1           3          4    754.0         735   1002.0   
2        2008      1           3          4    628.0         620    804.0   
3        2008      1           3          4   1829.0        1755   1959.0   
4        2008      1           3          4   1940.0        1915   2121.0   
...       ...    ...         ...        ...      ...         ...      ...   
1936753  2008     12          13          6   1250.0        1220   1617.0   
1936754  2008     12          13          6    657.0         600    904.0   
1936755  2008     12          13          6   1007.0         847   1149.0   
1936756  2008     12          13          6   1251.0        1240   1446.0   
1936757  2008     12          13          6   1110.0        1103   1413.0   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0              2225            WN        335  N712SW              128.0   
1              1000            WN       3231  N772SW              128.0   
2               750            WN        448  N428WN               96.0   
3              1925            WN       3920  N464WN               90.0   
4              2110            WN        378  N726SW              101.0   
...             ...           ...        ...     ...                ...   
1936753        1552            DL       1621  N938DL              147.0   
1936754         749            DL       1631  N3743H              127.0   
1936755        1010            DL       1631  N909DA              162.0   
1936756        1437            DL       1639  N646DL              115.0   
1936757        1418            DL       1641  N908DL              123.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
0                 150.0    116.0     -14.0       8.0    IAD  TPA       810   
1                 145.0    113.0       2.0      19.0    IAD  TPA       810   
2                  90.0     76.0      14.0       8.0    IND  BWI       515   
3                  90.0     77.0      34.0      34.0    IND  BWI       515   
4                 115.0     87.0      11.0      25.0    IND  JAX       688   
...                 ...      ...       ...       ...    ...  ...       ...   
1936753           152.0    120.0      25.0      30.0    MSP  ATL       906   
1936754           109.0     78.0      75.0      57.0    RIC  ATL       481   
1936755           143.0    122.0      99.0      80.0    ATL  IAH       689   
1936756           117.0     89.0       9.0      11.0    IAD  ATL       533   
1936757           135.0    104.0      -5.0       7.0    SAT  ATL       874   

         TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  \
0           4.0      8.0          0                N         0           NaN   
1           5.0     10.0          0                N         0           NaN   
2           3.0     17.0          0                N         0           NaN   
3           3.0     10.0          0                N         0           2.0   
4           4.0     10.0          0                N         0           NaN   
...         ...      ...        ...              ...       ...           ...   
1936753     9.0     18.0          0                N         0           3.0   
1936754    15.0     34.0          0                N         0           0.0   
1936755     8.0     32.0          0                N         0           1.0   
1936756    13.0     13.0          0                N         0           NaN   
1936757     8.0     11.0          0                N         0           NaN   

         WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
0                 NaN       NaN            NaN                NaN  
1                 NaN       NaN            NaN                NaN  
2                 NaN       NaN            NaN                NaN  
3            

## 2. Handling Null Values

### Gathering Initial Data About Null Values

In [7]:
# Determining number of null values in each column
print("Null records per column:")
total_nulls = df.isnull().sum()
total_nulls

Null records per column:


Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

In [8]:
# Determining the percentage of null values in each column
print("Percentage of null values per column:")
total_nulls / df.shape[0] * 100

Percentage of null values per column:


Year                  0.000000
Month                 0.000000
DayofMonth            0.000000
DayOfWeek             0.000000
DepTime               0.000000
CRSDepTime            0.000000
ArrTime               0.367108
CRSArrTime            0.000000
UniqueCarrier         0.000000
FlightNum             0.000000
TailNum               0.000258
ActualElapsedTime     0.433043
CRSElapsedTime        0.010223
AirTime               0.433043
ArrDelay              0.433043
DepDelay              0.000000
Origin                0.000000
Dest                  0.000000
Distance              0.000000
TaxiIn                0.367108
TaxiOut               0.023493
Cancelled             0.000000
CancellationCode      0.000000
Diverted              0.000000
CarrierDelay         35.588855
WeatherDelay         35.588855
NASDelay             35.588855
SecurityDelay        35.588855
LateAircraftDelay    35.588855
dtype: float64

### Dropping rows that have a low percentage of null values



The percentage of rows with null values for CRSElapsedTime is 0.010223%. Furthermore, the percentage of rows with null values for TailNum is 0.000258%.

Therefore, dataset is large enough to tolerate the loss of this data.

In [9]:
# Drop rows where CRSElapsedTime is null
df = df.dropna(subset=['CRSElapsedTime'])

# Drop rows where TailNum is null
df = df.dropna(subset=['TailNum'])

# Determining the percentage of null values in each column
print("Percentage of null values per column:")
df.isnull().sum() / df.shape[0] * 100

Percentage of null values per column:


Year                  0.000000
Month                 0.000000
DayofMonth            0.000000
DayOfWeek             0.000000
DepTime               0.000000
CRSDepTime            0.000000
ArrTime               0.356819
CRSArrTime            0.000000
UniqueCarrier         0.000000
FlightNum             0.000000
TailNum               0.000000
ActualElapsedTime     0.422761
CRSElapsedTime        0.000000
AirTime               0.422761
ArrDelay              0.422761
DepDelay              0.000000
Origin                0.000000
Dest                  0.000000
Distance              0.000000
TaxiIn                0.356819
TaxiOut               0.023495
Cancelled             0.000000
CancellationCode      0.000000
Diverted              0.000000
CarrierDelay         35.582207
WeatherDelay         35.582207
NASDelay             35.582207
SecurityDelay        35.582207
LateAircraftDelay    35.582207
dtype: float64

### Replacing null-valued rows with similar data





ActualElapsedTime describes flight duration, while CRSElapsedTime describes estimated flight duration. As a means of handling null ActualElapsedTime values, we replace them with the corresponding values from the CRSElapsedTime column.

ArrTime describes arrival time, while CRSArrTime describes estimated arrival time according CRS (central reservation systems). As a means of handling null ArrTime values, we replace them with the corresponding values from the CRSArrTime column.


In [10]:
# Replace ActualElapsedTime with CRSElapsedTime where ActualElapsedTime is null
ActualElapsedTime_null = df['ActualElapsedTime'].isnull()
df.loc[ActualElapsedTime_null, 'ActualElapsedTime'] = df.loc[ActualElapsedTime_null, 'CRSElapsedTime']

# Replace ArrTime with CRSArrTime where ArrTime is null
ArrTime_null = df['ArrTime'].isnull()
df.loc[ArrTime_null, 'ArrTime'] = df.loc[ArrTime_null, 'CRSArrTime']

# Determining the percentage of null values in each column
print("Percentage of null values per column:")
df.isnull().sum() / df.shape[0] * 100

Percentage of null values per column:


Year                  0.000000
Month                 0.000000
DayofMonth            0.000000
DayOfWeek             0.000000
DepTime               0.000000
CRSDepTime            0.000000
ArrTime               0.000000
CRSArrTime            0.000000
UniqueCarrier         0.000000
FlightNum             0.000000
TailNum               0.000000
ActualElapsedTime     0.000000
CRSElapsedTime        0.000000
AirTime               0.422761
ArrDelay              0.422761
DepDelay              0.000000
Origin                0.000000
Dest                  0.000000
Distance              0.000000
TaxiIn                0.356819
TaxiOut               0.023495
Cancelled             0.000000
CancellationCode      0.000000
Diverted              0.000000
CarrierDelay         35.582207
WeatherDelay         35.582207
NASDelay             35.582207
SecurityDelay        35.582207
LateAircraftDelay    35.582207
dtype: float64

### Handling attributes with null values within the same rows
If multiple attributes have the same percentage of null values, verify whether these null values occur within the same rows. If they do, proceed to drop those rows.

#### Handling "AirTime" and "ArrDelay" Null Values

Verify that "AirTime" and "ArrDelay" null values occur within the same rows

In [11]:
both_null = df['AirTime'].isnull() & df['ArrDelay'].isnull()
result_df = df[both_null]
result_size = result_df.shape[0]
total_rows = df.shape[0]
percentage = (result_size / total_rows) * 100
print("Percentage of df[both_null] relative to total rows:", percentage)

Percentage of df[both_null] relative to total rows: 0.4227610369961091


It is confirmed that null values for both 'AirTime' and 'ArrDelay' occur within the same rows. This is evident as the percentage of rows where both 'AirTime' and 'ArrDelay' are null is equal to the percentage of rows where each of them is null individually.

Thus, we may drop all rows where both ArrTime and ArrDelay are null.

In [12]:
df = df[~both_null]
# Determining the percentage of null values in each column
print("Percentage of null values per column:")
df.isnull().sum() / df.shape[0] * 100

Percentage of null values per column:


Year                  0.000000
Month                 0.000000
DayofMonth            0.000000
DayOfWeek             0.000000
DepTime               0.000000
CRSDepTime            0.000000
ArrTime               0.000000
CRSArrTime            0.000000
UniqueCarrier         0.000000
FlightNum             0.000000
TailNum               0.000000
ActualElapsedTime     0.000000
CRSElapsedTime        0.000000
AirTime               0.000000
ArrDelay              0.000000
DepDelay              0.000000
Origin                0.000000
Dest                  0.000000
Distance              0.000000
TaxiIn                0.000000
TaxiOut               0.000000
Cancelled             0.000000
CancellationCode      0.000000
Diverted              0.000000
CarrierDelay         35.308717
WeatherDelay         35.308717
NASDelay             35.308717
SecurityDelay        35.308717
LateAircraftDelay    35.308717
dtype: float64

#### Handling "CarrierDelay", "WeatherDelay", "NASDelay", "SecurityDelay", and "LateAircraftDelay" Null Values

Verify that "CarrierDelay", "WeatherDelay", "NASDelay", "SecurityDelay", and "LateAircraftDelay" null values occur within the same rows

In [13]:
all_null = df['CarrierDelay'].isnull() & df['WeatherDelay'].isnull() & df['NASDelay'].isnull() & df['SecurityDelay'].isnull() & df['LateAircraftDelay'].isnull()
result_df = df[all_null]
result_size = result_df.shape[0]
total_rows = df.shape[0]
percentage = (result_size / total_rows) * 100
print("Percentage of df[all_null] relative to total rows:", percentage)

Percentage of df[all_null] relative to total rows: 35.30871700837185


It is confirmed that null values for "CarrierDelay", "WeatherDelay", "NASDelay", "SecurityDelay", and "LateAircraftDelay" occur within the same rows. This is evident as the percentage of rows where "CarrierDelay", "WeatherDelay", "NASDelay", "SecurityDelay", and "LateAircraftDelay" are null is equal to the percentage of rows where each of them is null individually.

Thus, we may drop all rows where "CarrierDelay", "WeatherDelay", "NASDelay", "SecurityDelay", and "LateAircraftDelay" are null.

In [14]:
df = df[~all_null]
# Determining the percentage of null values in each column
print("Percentage of null values per column:")
df.isnull().sum() / df.shape[0] * 100

Percentage of null values per column:


Year                 0.0
Month                0.0
DayofMonth           0.0
DayOfWeek            0.0
DepTime              0.0
CRSDepTime           0.0
ArrTime              0.0
CRSArrTime           0.0
UniqueCarrier        0.0
FlightNum            0.0
TailNum              0.0
ActualElapsedTime    0.0
CRSElapsedTime       0.0
AirTime              0.0
ArrDelay             0.0
DepDelay             0.0
Origin               0.0
Dest                 0.0
Distance             0.0
TaxiIn               0.0
TaxiOut              0.0
Cancelled            0.0
CancellationCode     0.0
Diverted             0.0
CarrierDelay         0.0
WeatherDelay         0.0
NASDelay             0.0
SecurityDelay        0.0
LateAircraftDelay    0.0
dtype: float64

## 3. Ensuring certain columns aren't redundant

Dropping the following three columns because they only have one unique value so they're not useful for any data anlytics

In [15]:
df['Cancelled'].value_counts()
df.drop(columns=['Cancelled'], inplace=True)

In [16]:
df['Diverted'].value_counts()
df.drop(columns=['Diverted'], inplace=True)

In [17]:
df['CancellationCode'].value_counts()
df.drop(columns=['CancellationCode'], inplace=True)

## 4. Identifying Columns with Repeating Values and Ensuring they Make Sense Considering the Context



In [18]:
# Checking all the unique flight carriers
df['UniqueCarrier'].value_counts()

WN    203559
AA    132257
MQ     97555
UA     95465
OO     88991
DL     72252
XE     72008
US     59508
CO     58958
EV     56781
NW     54744
YV     50646
FL     46991
OH     39292
B6     38232
9E     35630
AS     24012
F9     15940
HA      4325
AQ       340
Name: UniqueCarrier, dtype: int64

In [19]:
# Checking all the unique flight numbers
df['FlightNum'].value_counts()

50      1053
321      930
75       922
16       915
40       902
        ... 
7712       1
7477       1
6249       1
6268       1
9741       1
Name: FlightNum, Length: 7481, dtype: int64

In [20]:
# Checking all the unique Tail numbers
df['TailNum'].value_counts()

N17175    686
N77278    679
N87353    678
N37208    665
N37342    660
         ... 
N853NW      1
N104UA      1
N78003      1
N118UA      1
N78008      1
Name: TailNum, Length: 5349, dtype: int64

# Data Transformation

## Converting Attributes to Correct Data Types

### Identifying Data Types in DF

In [21]:
print("Data Types:")
print(df.dtypes)

Data Types:
Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn               float64
TaxiOut              float64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object


### Converting all 'object' Types to 'string' Types (where appropriate)

As per the [Pandas Documentation](https://pandas.pydata.org/docs/user_guide/text.html),  it is recommended to use the StringDtype rather than the object dtype due to clarity and consistency.

In [22]:
df['UniqueCarrier'] = df['UniqueCarrier'].astype("string")
df['TailNum'] = df['TailNum'].astype("string")
df['Origin'] = df['Origin'].astype("string")
df['Dest'] = df['Dest'].astype("string")

print("Data Types:\n")
print(df.dtypes)

Data Types:

Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         string
FlightNum              int64
TailNum               string
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                string
Dest                  string
Distance               int64
TaxiIn               float64
TaxiOut              float64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object


### Unifying Arrival and Dept Times by Converting them to ints

In [23]:
df['ArrTime'] = df['ArrTime'].astype(int)
df['DepTime'] = df['DepTime'].astype(int)

print("Data Types:\n")
print(df.dtypes)

Data Types:

Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime                int64
CRSDepTime             int64
ArrTime                int64
CRSArrTime             int64
UniqueCarrier         string
FlightNum              int64
TailNum               string
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                string
Dest                  string
Distance               int64
TaxiIn               float64
TaxiOut              float64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object


### Converting HHMM to Timestamp

In [25]:
# Timestamps that are 2400 should be converted to 0000 since that represents 12am
df[df['ArrTime'].astype(str) == '2400']

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1281     2008      1           3          4     2238        2150     2400   
2201     2008      1           4          5     1844        1815     2400   
4057     2008      1           5          6     2236        1805     2400   
6028     2008      1           6          7     2153        2055     2400   
6426     2008      1           6          7     2212        2105     2400   
...       ...    ...         ...        ...      ...         ...      ...   
1918886  2008     12          21          7     2003        1459     2400   
1920920  2008     12          28          7     2045        2005     2400   
1921381  2008     12          30          2     2238        2150     2400   
1921603  2008     12          31          3     2037        2000     2400   
1932497  2008     12           1          1     2225        2120     2400   

         CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
1281           2320            WN       1320  N456WN               82.0   
2201           2335            WN       2450  N700GS              196.0   
4057           1930            WN        646  N283WN               84.0   
6028           2255            WN       2730  N900WN              127.0   
6426           2225            WN       1482  N675AA              108.0   
...             ...           ...        ...     ...                ...   
1918886        1813            B6          9  N639JB              237.0   
1920920        2344            B6       1069  N178JB              255.0   
1921381        2328            B6       1079  N258JB               82.0   
1921603        2312            B6         43  N561JB              203.0   
1932497        2304            DL        830  N969DL               95.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
1281               90.0     70.0      40.0      48.0    SAN  SMF       480   
2201              200.0    177.0      25.0      29.0    LAS  MDW      1521   
4057               85.0     71.0     270.0     271.0    LAX  SFO       337   
6028              120.0    104.0      65.0      58.0    OAK  SEA       671   
6426               80.0     74.0      95.0      67.0    SAN  SJC       417   
...                 ...      ...       ...       ...    ...  ...       ...   
1918886           194.0    162.0     347.0     304.0    JFK  FLL      1069   
1920920           279.0    233.0      16.0      40.0    JFK  AUS      1522   
1921381            98.0     57.0      32.0      48.0    JFK  RIC       288   
1921603           192.0    139.0      48.0      37.0    JFK  MCO       944   
1932497           104.0     71.0      56.0      65.0    ATL  DCA       547   

         TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1281        4.0      8.0           1.0           0.0       0.0            0.0   
2201        6.0     13.0           7.0           0.0       0.0            0.0   
4057        6.0      7.0         105.0           0.0       0.0            0.0   
6028        4.0     19.0          21.0           0.0       7.0            0.0   
6426        6.0     28.0           0.0          36.0      28.0            0.0   
...         ...      ...           ...           ...       ...            ...   
1918886     5.0     70.0         146.0           0.0      43.0            0.0   
1920920     4.0     18.0          16.0           0.0       0.0            0.0   
1921381     2.0     23.0           1.0           0.0       0.0            0.0   
1921603     4.0     60.0          37.0           0.0      11.0            0.0   
1932497     7.0     17.0           0.0           0.0       0.0            0.0   

         LateAircraftDelay  
1281                  39.0  
2201                  18.0  
4057                 165.0  
6028                  37.0  
6426                  31.0  
...                    ...  
1918886              158.0  
1920920                0.0  
1921381     

In [26]:
from datetime import time

# Convert from int to string to padd with leading zeros
df['DepTime'] = df['DepTime'].astype(str).str.zfill(4)
df['ArrTime'] = df['ArrTime'].astype(str).str.zfill(4)
df['CRSDepTime'] = df['CRSDepTime'].astype(str).str.zfill(4)
df['CRSArrTime'] = df['CRSArrTime'].astype(str).str.zfill(4)


def convert_to_timestamp(time_str):
  if time_str == '2400':
    time_str = '0000'
  return time(int(time_str[:2]), int(time_str[2:]))

# Applying the conversion function to all rows of the 'time_column' column
df['DepTime'] = df['DepTime'].apply(lambda x: convert_to_timestamp(x))
df['ArrTime'] = df['ArrTime'].apply(lambda x: convert_to_timestamp(x))
df['CRSDepTime'] = df['CRSDepTime'].apply(lambda x: convert_to_timestamp(x))
df['CRSArrTime'] = df['CRSArrTime'].apply(lambda x: convert_to_timestamp(x))

### Converting FlightNum to String

In [27]:
df['FlightNum'] = df['FlightNum'].astype("string")

print("Data Types:\n")
print(df.dtypes)

print("\nSample data with Timestamps\n")
df[:10]

Data Types:

Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime               object
CRSDepTime            object
ArrTime               object
CRSArrTime            object
UniqueCarrier         string
FlightNum             string
TailNum               string
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                string
Dest                  string
Distance               int64
TaxiIn               float64
TaxiOut              float64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object

Sample data with Timestamps



Year  Month  DayofMonth  DayOfWeek   DepTime CRSDepTime   ArrTime  \
3   2008      1           3          4  18:29:00   17:55:00  19:59:00   
5   2008      1           3          4  19:37:00   18:30:00  20:37:00   
7   2008      1           3          4  16:44:00   15:10:00  18:45:00   
9   2008      1           3          4  14:52:00   14:25:00  16:40:00   
11  2008      1           3          4  13:23:00   12:55:00  15:26:00   
12  2008      1           3          4  14:16:00   13:25:00  15:12:00   
13  2008      1           3          4  16:57:00   16:25:00  17:54:00   
17  2008      1           3          4  14:22:00   12:55:00  16:57:00   
19  2008      1           3          4  21:07:00   19:45:00  23:34:00   
23  2008      1           3          4  18:12:00   16:50:00  19:27:00   

   CRSArrTime UniqueCarrier FlightNum TailNum  ActualElapsedTime  \
3    19:25:00            WN      3920  N464WN               90.0   
5    19:40:00            WN       509  N763SW              240.0   
7    17:25:00            WN      1333  N334SW              121.0   
9    16:25:00            WN       675  N286WN              228.0   
11   15:10:00            WN         4  N674AA              123.0   
12   14:35:00            WN        54  N643SW               56.0   
13   17:35:00            WN       623  N724SW               57.0   
17   16:10:00            WN       188  N215WN              155.0   
19   22:30:00            WN       362  N798SW              147.0   
23   18:15:00            WN       422  N779SW              135.0   

    CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
3             90.0     77.0      34.0      34.0    IND  BWI       515     3.0   
5            250.0    230.0      57.0      67.0    IND  LAS      1591     3.0   
7            135.0    107.0      80.0      94.0    IND  MCO       828     6.0   
9            240.0    213.0      15.0      27.0    IND  PHX      1489     7.0   
11           135.0    110.0      16.0      28.0    IND  TPA       838     4.0   
12            70.0     49.0      37.0      51.0    ISP  BWI       220     2.0   
13            70.0     47.0      19.0      32.0    ISP  BWI       220     5.0   
17           195.0    143.0      47.0      87.0    ISP  FLL      1093     6.0   
19           165.0    134.0      64.0      82.0    ISP  MCO       972     6.0   
23           145.0    118.0      72.0      82.0    ISP  MDW       765     6.0   

    TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
3      10.0           2.0           0.0       0.0            0.0   
5       7.0          10.0           0.0       0.0            0.0   
7       8.0           8.0           0.0       0.0            0.0   
9       8.0           3.0           0.0       0.0            0.0   
11      9.0           0.0           0.0       0.0            0.0   
12      5.0          12.0           0.0       0.0            0.0   
13      5.0           7.0           0.0       0.0            0.0   
17      6.0          40.0           0.0       0.0            0.0   
19      7.0           5.0           0.0       0.0            0.0   
23     11.0           3.0           0.0       0.0            0.0   

    LateAircraftDelay  
3                32.0  
5                47.0  
7                72.0  
9                12.0  
11               16.0  
12               25.0  
13               12.0  
17                7.0  
19               59.0  
23               69.0

# Data Validation


## Checking unique values to see if they comply with our conceptual model

In [28]:
def get_unique(column):
    return column.unique()

df.apply(get_unique, axis=0)

Year                                                            [2008]
Month                          [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
DayofMonth           [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, ...
DayOfWeek                                        [4, 5, 6, 7, 1, 2, 3]
DepTime              [18:29:00, 19:37:00, 16:44:00, 14:52:00, 13:23...
CRSDepTime           [17:55:00, 18:30:00, 15:10:00, 14:25:00, 12:55...
ArrTime              [19:59:00, 20:37:00, 18:45:00, 16:40:00, 15:26...
CRSArrTime           [19:25:00, 19:40:00, 17:25:00, 16:25:00, 15:10...
UniqueCarrier        [WN, XE, YV, OH, OO, UA, US, DL, EV, F9, FL, H...
FlightNum            [3920, 509, 1333, 675, 4, 54, 623, 188, 362, 4...
TailNum              [N464WN, N763SW, N334SW, N286WN, N674AA, N643S...
ActualElapsedTime    [90.0, 240.0, 121.0, 228.0, 123.0, 56.0, 57.0,...
CRSElapsedTime       [90.0, 250.0, 135.0, 240.0, 70.0, 195.0, 165.0...
AirTime              [77.0, 230.0, 107.0, 213.0, 110.0, 49.0, 47.0,...
ArrDel

## Obtaining statistical information about the DF

In [29]:
df.describe()

Year         Month    DayofMonth     DayOfWeek  ActualElapsedTime  \
count  1247486.0  1.247486e+06  1.247486e+06  1.247486e+06       1.247486e+06   
mean      2008.0  6.065399e+00  1.572542e+01  3.980082e+00       1.353779e+02   
std          0.0  3.508937e+00  8.793008e+00  1.993270e+00       7.229636e+01   
min       2008.0  1.000000e+00  1.000000e+00  1.000000e+00       1.400000e+01   
25%       2008.0  3.000000e+00  8.000000e+00  2.000000e+00       8.300000e+01   
50%       2008.0  6.000000e+00  1.600000e+01  4.000000e+00       1.180000e+02   
75%       2008.0  9.000000e+00  2.300000e+01  6.000000e+00       1.670000e+02   
max       2008.0  1.200000e+01  3.100000e+01  7.000000e+00       1.114000e+03   

       CRSElapsedTime       AirTime      ArrDelay      DepDelay      Distance  \
count    1.247486e+06  1.247486e+06  1.247486e+06  1.247486e+06  1.247486e+06   
mean     1.317641e+02  1.074203e+02  6.329104e+01  5.967721e+01  7.415867e+02   
std      6.953201e+01  6.817454e+01  6.075380e+01  5.986881e+01  5.593643e+02   
min     -2.100000e+01  0.000000e+00  1.500000e+01  6.000000e+00  1.100000e+01   
25%      8.000000e+01  5.800000e+01  2.600000e+01  2.400000e+01  3.340000e+02   
50%      1.150000e+02  9.000000e+01  4.300000e+01  4.100000e+01  5.950000e+02   
75%      1.610000e+02  1.360000e+02  7.900000e+01  7.500000e+01  9.720000e+02   
max      6.600000e+02  1.091000e+03  2.461000e+03  2.467000e+03  4.962000e+03   

             TaxiIn       TaxiOut  CarrierDelay  WeatherDelay      NASDelay  \
count  1.247486e+06  1.247486e+06  1.247486e+06  1.247486e+06  1.247486e+06   
mean   7.297232e+00  2.066033e+01  1.917943e+01  3.703355e+00  1.502162e+01   
std    6.033239e+00  1.667983e+01  4.354624e+01  2.149153e+01  3.383308e+01   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    4.000000e+00  1.100000e+01  0.000000e+00  0.000000e+00  0.000000e+00   
50%    6.000000e+00  1.600000e+01  2.000000e+00  0.000000e+00  2.000000e+00   
75%    8.000000e+00  2.400000e+01  2.100000e+01  0.000000e+00  1.500000e+01   
max    2.400000e+02  4.220000e+02  2.436000e+03  1.352000e+03  1.357000e+03   

       SecurityDelay  LateAircraftDelay  
count   1.247486e+06       1.247486e+06  
mean    9.013728e-02       2.529649e+01  
std     2.022716e+00       4.205489e+01  
min     0.000000e+00       0.000000e+00  
25%     0.000000e+00       0.000000e+00  
50%     0.000000e+00       8.000000e+00  
75%     0.000000e+00       3.300000e+01  
max     3.920000e+02       1.316000e+03

## Finding the range of values for each comumn in the DF and ensuring they comply with our conceptual model

In [30]:
df.describe().loc[['min', 'max']]

Year  Month  DayofMonth  DayOfWeek  ActualElapsedTime  CRSElapsedTime  \
min  2008.0    1.0         1.0        1.0               14.0           -21.0   
max  2008.0   12.0        31.0        7.0             1114.0           660.0   

     AirTime  ArrDelay  DepDelay  Distance  TaxiIn  TaxiOut  CarrierDelay  \
min      0.0      15.0       6.0      11.0     0.0      0.0           0.0   
max   1091.0    2461.0    2467.0    4962.0   240.0    422.0        2436.0   

     WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
min           0.0       0.0            0.0                0.0  
max        1352.0    1357.0          392.0             1316.0

# Data Loading (splitting into different data frames)


## Date data frame

In [31]:
date_df = df.iloc[:, :4]
date_df.drop_duplicates(subset=['Year','Month','DayofMonth','DayOfWeek'], inplace=True)
date_df = date_df.sort_values(['Month','DayofMonth'], ascending=[True, True])
date_df

Year  Month  DayofMonth  DayOfWeek
30743    2008      1           1          2
30736    2008      1           2          3
3        2008      1           3          4
1661     2008      1           4          5
3409     2008      1           5          6
...       ...    ...         ...        ...
1765779  2008     12          27          6
1767819  2008     12          28          7
1769247  2008     12          29          1
1770674  2008     12          30          2
1772089  2008     12          31          3

[366 rows x 4 columns]

## Location data frame

In [32]:
# location_df = df.iloc[:, 16:18]

union_values = sorted(set(df['Origin']).union(set(df['Dest'])))
location_df = pd.DataFrame({'Airport_Code': list(union_values)})
location_df

Airport_Code
0            ABE
1            ABI
2            ABQ
3            ABY
4            ACK
..           ...
297          WYS
298          XNA
299          YAK
300          YKM
301          YUM

[302 rows x 1 columns]

## Aircraft data frame

In [33]:
aircraft_df = df.iloc[:, 8:11]
aircraft_df.drop(columns=["FlightNum"], inplace = True)
aircraft_df.drop_duplicates(subset=['UniqueCarrier','TailNum'], inplace=True)
aircraft_df[aircraft_df['TailNum'] == 'N464WN']

UniqueCarrier TailNum
3            WN  N464WN

# Surrogate Key Pipeline

In [34]:
# Reset the index after dropping previous rows
df = df.reset_index(drop=True)
df

Year  Month  DayofMonth  DayOfWeek   DepTime CRSDepTime   ArrTime  \
0        2008      1           3          4  18:29:00   17:55:00  19:59:00   
1        2008      1           3          4  19:37:00   18:30:00  20:37:00   
2        2008      1           3          4  16:44:00   15:10:00  18:45:00   
3        2008      1           3          4  14:52:00   14:25:00  16:40:00   
4        2008      1           3          4  13:23:00   12:55:00  15:26:00   
...       ...    ...         ...        ...       ...        ...       ...   
1247481  2008     12          13          6  09:21:00   08:30:00  11:12:00   
1247482  2008     12          13          6  15:52:00   15:20:00  17:35:00   
1247483  2008     12          13          6  12:50:00   12:20:00  16:17:00   
1247484  2008     12          13          6  06:57:00   06:00:00  09:04:00   
1247485  2008     12          13          6  10:07:00   08:47:00  11:49:00   

        CRSArrTime UniqueCarrier FlightNum TailNum  ActualElapsedTime  \
0         19:25:00            WN      3920  N464WN               90.0   
1         19:40:00            WN       509  N763SW              240.0   
2         17:25:00            WN      1333  N334SW              121.0   
3         16:25:00            WN       675  N286WN              228.0   
4         15:10:00            WN         4  N674AA              123.0   
...            ...           ...       ...     ...                ...   
1247481   10:08:00            DL      1616  N907DE              111.0   
1247482   17:18:00            DL      1620  N905DE               43.0   
1247483   15:52:00            DL      1621  N938DL              147.0   
1247484   07:49:00            DL      1631  N3743H              127.0   
1247485   10:10:00            DL      1631  N909DA              162.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
0                  90.0     77.0      34.0      34.0    IND  BWI       515   
1                 250.0    230.0      57.0      67.0    IND  LAS      1591   
2                 135.0    107.0      80.0      94.0    IND  MCO       828   
3                 240.0    213.0      15.0      27.0    IND  PHX      1489   
4                 135.0    110.0      16.0      28.0    IND  TPA       838   
...                 ...      ...       ...       ...    ...  ...       ...   
1247481            98.0     82.0      64.0      51.0    ATL  PBI       545   
1247482            58.0     27.0      17.0      32.0    HSV  ATL       151   
1247483           152.0    120.0      25.0      30.0    MSP  ATL       906   
1247484           109.0     78.0      75.0      57.0    RIC  ATL       481   
1247485           143.0    122.0      99.0      80.0    ATL  IAH       689   

         TaxiIn  TaxiOut  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0           3.0     10.0           2.0           0.0       0.0            0.0   
1           3.0      7.0          10.0           0.0       0.0            0.0   
2           6.0      8.0           8.0           0.0       0.0            0.0   
3           7.0      8.0           3.0           0.0       0.0            0.0   
4           4.0      9.0           0.0           0.0       0.0            0.0   
...         ...      ...           ...           ...       ...            ...   
1247481     8.0     21.0          51.0           0.0      13.0            0.0   
1247482     9.0      7.0           0.0           0.0       0.0            0.0   
1247483     9.0     18.0           3.0           0.0       0.0            0.0   
1247484    15.0     34.0           0.0          57.0      18.0            0.0   
1247485     8.0     32.0           1.0           0.0      19.0            0.0   

         LateAircraftDelay  
0                     32.0  
1                     47.0  
2                     72.0  
3                     12.0  
4                     16.0  
...                    ...  
1247481                0.0  
1247482               17.0  
1247483               22

## Aircraft Dimension Surrogate key

In [35]:
# Surrogate key generation for Aircraft dimension
aircraft_df = aircraft_df.reset_index(drop=True)

aircraft_df['Aircraft_Key'] = range(1,len(aircraft_df)+1)
aircraft_df = aircraft_df.reindex(columns=['Aircraft_Key'] + list([c for c in aircraft_df.columns if c!= 'Aircraft_Key']))
aircraft_df

Aircraft_Key UniqueCarrier TailNum
0                1            WN  N464WN
1                2            WN  N763SW
2                3            WN  N334SW
3                4            WN  N286WN
4                5            WN  N674AA
...            ...           ...     ...
5346          5347            B6  N763JB
5347          5348            B6  N766JB
5348          5349            CO  N75428
5349          5350            CO  N75429
5350          5351            CO  N78008

[5351 rows x 3 columns]

## Date Dimension Surrogate key

In [36]:
# Surrogate key generation for Date dimension
date_df = date_df.reset_index(drop=True)

date_df['Date_Key'] = range(1,len(date_df)+1)
date_df = date_df.reindex(columns=['Date_Key'] + list([c for c in date_df.columns if c!= 'Date_Key']))
date_df

Date_Key  Year  Month  DayofMonth  DayOfWeek
0           1  2008      1           1          2
1           2  2008      1           2          3
2           3  2008      1           3          4
3           4  2008      1           4          5
4           5  2008      1           5          6
..        ...   ...    ...         ...        ...
361       362  2008     12          27          6
362       363  2008     12          28          7
363       364  2008     12          29          1
364       365  2008     12          30          2
365       366  2008     12          31          3

[366 rows x 5 columns]

## Location Dimension Surrogate key

In [37]:
# Surrogate key generation for Location dimension
location_df = location_df.reset_index(drop=True)

location_df['Location_Key'] = range(1,len(location_df)+1)
location_df = location_df.reindex(columns=['Location_Key'] + list([c for c in location_df.columns if c!= 'Location_Key']))
location_df

Location_Key Airport_Code
0               1          ABE
1               2          ABI
2               3          ABQ
3               4          ABY
4               5          ACK
..            ...          ...
297           298          WYS
298           299          XNA
299           300          YAK
300           301          YKM
301           302          YUM

[302 rows x 2 columns]

# Fact Table Staging


## Location Dimension Staging

In [38]:
main_df = df.copy(deep=True)
main_df = main_df.merge(location_df, how='left', left_on='Origin', right_on='Airport_Code')
main_df.drop(columns=['Origin', 'Airport_Code'], inplace=True)
main_df.rename(columns={'Location_Key': 'Origin_Key'}, inplace=True)

main_df = main_df.merge(location_df, how='left', left_on='Dest', right_on='Airport_Code')
main_df.drop(columns=['Dest', 'Airport_Code'], inplace=True)
main_df.rename(columns={'Location_Key': 'Dest_Key'}, inplace=True)

## Date Dimension Staging

In [39]:


main_df = main_df.merge(date_df, how='left', on=['Year', 'Month', 'DayofMonth', 'DayOfWeek'])
main_df.drop(columns=['Year', 'Month', 'DayofMonth', 'DayOfWeek'], inplace=True)
main_df.rename(columns={'Date_Key': 'Date_Key'}, inplace=True)


## Aircraft Dimension Staging

In [40]:


main_df = main_df.merge(aircraft_df, how='left', on=['UniqueCarrier', 'TailNum'])
main_df.drop(columns=['UniqueCarrier', 'TailNum'], inplace=True)
main_df.rename(columns={'Aircraft_Key': 'Aircraft_Key'}, inplace=True)

## Aggregation of Total Delay Time

In [41]:
main_df['TotalDelayTime'] = main_df[['CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay', 'DepDelay']].sum(axis=1)

# Final Fact Table

In [42]:
main_df = main_df[['Date_Key', 'Aircraft_Key', 'Origin_Key', 'Dest_Key', 'FlightNum',
       'Distance', 'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay', 'ArrDelay',
       'DepDelay', 'TaxiIn', 'TaxiOut', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'DepTime','CRSDepTime',
       'ArrTime', 'CRSArrTime', 'TotalDelayTime'
          ]]

main_df = main_df.sort_values(['Date_Key'], ascending=[True])
main_df

Date_Key  Aircraft_Key  Origin_Key  Dest_Key FlightNum  Distance  \
59779           1           212          49        36      2934       588   
70275           1          2653          81       221       851       602   
70276           1          2653         221        81       424       602   
70277           1          2653          81       156       775       629   
70278           1          2653         156        81       770       629   
...           ...           ...         ...       ...       ...       ...   
1151346       366          1232         260       237      5482       199   
1173833       366          1839          97        64       687       529   
1173832       366          1886          88        64      1682       500   
1173820       366          1972          64       289      1851      1735   
1196718       366          2685         188       156       216      1524   

         CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
59779            11.0           0.0       0.0            0.0   
70275            25.0           0.0      11.0            0.0   
70276             0.0           0.0      19.0            0.0   
70277            14.0           0.0       3.0            0.0   
70278             3.0           0.0      17.0            0.0   
...               ...           ...       ...            ...   
1151346           0.0           0.0       0.0            0.0   
1173833          13.0           0.0       0.0            0.0   
1173832          14.0           0.0       1.0            0.0   
1173820          10.0           0.0      10.0            0.0   
1196718          21.0           0.0       4.0            0.0   

         LateAircraftDelay  ArrDelay  DepDelay  TaxiIn  TaxiOut  \
59779                  4.0      15.0      17.0     4.0     10.0   
70275                  0.0      36.0      47.0     4.0     11.0   
70276                 25.0      44.0      32.0     8.0     25.0   
70277                  0.0      17.0      16.0    10.0      8.0   
70278                  2.0      22.0      19.0     7.0     19.0   
...                    ...       ...       ...     ...      ...   
1151346               56.0      56.0      56.0     5.0     15.0   
1173833               30.0      43.0      59.0     8.0     20.0   
1173832                3.0      18.0      17.0     6.0     15.0   
1173820                3.0      23.0      13.0     4.0     27.0   
1196718                0.0      25.0      21.0     6.0     12.0   

         ActualElapsedTime  CRSElapsedTime  AirTime   DepTime CRSDepTime  \
59779                118.0           120.0    104.0  09:42:00   09:25:00   
70275                106.0           117.0     91.0  09:07:00   08:20:00   
70276                118.0           106.0     85.0  11:29:00   10:57:00   
70277                115.0           114.0     97.0  14:21:00   14:05:00   
70278                113.0           110.0     87.0  15:59:00   15:40:00   
...                    ...             ...      ...       ...        ...   
1151346               65.0            65.0     45.0  14:26:00   13:30:00   
1173833              110.0           126.0     82.0  18:09:00   17:10:00   
1173832              102.0           101.0     81.0  19:17:00   19:00:00   
1173820              276.0           266.0    245.0  20:23:00   20:10:00   
1196718              237.0           233.0    219.0  09:16:00   08:55:00   

          ArrTime CRSArrTime  TotalDelayTime  
59779    10:40:00   10:25:00            32.0  
70275    10:53:00   10:17:00            83.0  
70276    13:27:00   12:43:00            76.0  
70277    15:16:00   14:59:00            33.0  
70278    18:52:00   18:30:00            41.0  
...           ...        ...             ...  
1151346  15:31:00   14:35:00           112.0  
1173833  19:59:00   19:16:00           102.0  
1173832  20:59:00   20:41:00            35.0  
1173820  22:59:00   22:36:00            36.0  
1196718  11:13:00   10:48:00            46.0  

[1247486 rows x 23 columns]

# Uploading Dataframes to SQL

In [43]:
!pip install python-dotenv
!pip install sqlalchemy

from sqlalchemy import create_engine
from dotenv import load_dotenv
import os

## Loading Env Variables

In [44]:
dotenv_path = './.env'
load_dotenv(dotenv_path)

dbname = os.getenv('DB_NAME')
user = os.getenv('DB_USER')
password = os.getenv('DB_PASS')
host = os.getenv('DB_HOST')
port = os.getenv('DB_PORT')

## Establishing Connection

In [45]:
try:
  engine = create_engine(f'postgresql://{user}:{password}@{host}:{port}/{dbname}')
  print("Database connected successfully")
except Exception as e:
  print("Error connecting to the database:", e)

Database connected successfully


## Creating SQL Tables

### Creating Location Dimension

In [46]:
try:
  location_df.to_sql('location_dimension', engine, index=False)
  print("Location DataFrame uploaded to PostgreSQL successfully.")
except Exception as e:
  print("Error uploading DataFrame to PostgreSQL:", e)

Location DataFrame uploaded to PostgreSQL successfully.


### Creating Aircraft Dimension

In [47]:
try:
  aircraft_df.to_sql('aircraft_dimension', engine, index=False)
  print("Aircraft DataFrame uploaded to PostgreSQL successfully.")
except Exception as e:
  print("Error uploading DataFrame to PostgreSQL:", e)

Aircraft DataFrame uploaded to PostgreSQL successfully.


### Creating Date Dimension

In [48]:
try:
  date_df.to_sql('date_dimension', engine, index=False)
  print("Date DataFrame uploaded to PostgreSQL successfully.")
except Exception as e:
  print("Error uploading DataFrame to PostgreSQL:", e)

Date DataFrame uploaded to PostgreSQL successfully.


### Creating Fact Table

In [49]:
try:
  main_df.to_sql('flights_fact_table', engine, index=False)
  print("Flights Fact Table DataFrame uploaded to PostgreSQL successfully.")
except Exception as e:
  print("Error uploading DataFrame to PostgreSQL:", e)

Flights Fact Table DataFrame uploaded to PostgreSQL successfully.


## Closing the PostgreSQL Connection

In [50]:
engine.dispose()